src/usdt/mod.rs

```diff
@@ -0,0 +1,187 @@
+    pid: Option<pid_t>,
+    probes: HashSet<(String, String, String)>,
+}
+
+impl USDTContext {
+    /// Create a new USDT context from a PID.
+    pub fn from_pid(pid: pid_t) -> Result<Self, BccError> {
+        Self::new(Some(pid), None)
+    }
+
+    /// Create a new USDT context from a path to the binary.
+    pub fn from_binary_path<T: AsRef<Path>>(path: T) -> Result<Self, BccError> {
+        Self::new(None, Some(PathBuf::from(path.as_ref())))
+    }
+
+    /// Create a new USDT context from a path to the binary and a specific PID.
+    pub fn from_binary_path_and_pid<T: AsRef<Path>>(path: T, pid: pid_t) -> Result<Self, BccError> {
+        Self::new(Some(pid), Some(PathBuf::from(path.as_ref())))
+    }
+
+    fn new(pid: Option<pid_t>, path: Option<PathBuf>) -> Result<Self, BccError> {
+        let context = match (pid, path) {
+            (None, None) => ptr::null_mut(),
+            (None, Some(path)) => {
+                let c_path = CString::new(path.as_os_str().as_bytes().to_owned())?;
+                unsafe { bcc_usdt_new_frompath(c_path.as_ptr()) }
+            },
+            (Some(pid), None) => {
+                unsafe { bcc_usdt_new_frompid(pid, ptr::null()) }
+            },
+            (Some(pid), Some(path)) => {
+                let c_path = CString::new(path.as_os_str().as_bytes().to_owned())?;
+                unsafe { bcc_usdt_new_frompid(pid, c_path.as_ptr()) } 
+            },
+        };
+
+        if context.is_null() {
+            Err(BccError::CreateUSDTContext { message: format!("failed to create USDT context") })
+        } else {
+            Ok(Self { context, pid, probes: HashSet::new() })
+        }
+    }
+
+    /// Enables a probe, calling a function in the BPF program whenever the tracepoint is hit.
+    ///
+    /// The `probe` argument is the name of tracepoint to enable.  The probe name can either be in
+    /// the generic form -- simply the name of the tracepoint itself -- or the prefixed form of
+    /// `<provider>:<name>`.
+    ///
+    /// The `fn_name` argument is the name of the function in the BPF program to call when the given
+    /// tracepoint is hit.  This function will be called with the given arguments for the tracepoint.
+    pub fn enable_probe(&mut self, probe: impl Into<String>, fn_name: impl Into<String>) -> Result<(), BccError> {
+        let probe = probe.into();
+        let fn_name = fn_name.into();
+
+        let mut probe_parts = probe.split(":").map(|s| s.to_owned()).collect::<Vec<_>>();
+        let fn_name_c = CString::new(fn_name.clone())?;
+        let result = if probe_parts.len() == 1 {
+            let probe_c = CString::new(probe.clone())?;
+            let result = unsafe { bcc_usdt_enable_probe(self.context, probe_c.as_ptr(), fn_name_c.as_ptr()) };
+            if result == 0 {
+                self.probes.insert((String::new(), probe.clone(), fn_name));
+            }
+            result
+        } else {
+            let provider = probe_parts.remove(0);
```

Comment:
Probably shouldn't happen, but maybe we don't want to panic if probe_parts is empty?

Reply:
I flipped it so that it's "if len == 2, use the provider/probe logic, otherwise, use the single string logic".